src/inthandler.S

```diff
@@ -226,9 +219,18 @@ endint:
 	ldc1 $f30,saveFR30
 	ldc1 $f31,saveFR31
 
-	ld $30,saveFC31
-	nop
+	lw $30,saveFC31
 
+	# Override the cause bits (12-17) of stored value from cop1.
+	# If they were cleared by the ex. handler, they will remain cleared.
+	# We will restore the other values to their state before the exception
+	# has occured. If we restore all the bits, the exception handler will
+	# be unable to clear them if needed. This is important as they will
+	# re-trigger the exception upon handler exit if remains set.
+	cfc1 k0, $f31
+	and k0, 0x3F000
+	or k0, ~0x3F000
+	and $30, k0
```

Comment:
OK this makes sense if we keep the current approach where the exception handler registered by the application has the task to also clear the exception (contrary to interrupts, where the interrupt is acknowledge by libdragon itself before calling the handler).

So I would maybe document at the top `exception.c` and in the comment of `register_exception_handler` that the handler must clear the exception in case it's required to avoid re-triggering.

Reply:
I think it is better to leave this to the consumer indeed. Floating point exceptions are disabled by default and user may never decide to enable them.

I initially had an alternative implementation where I removed the const specifier from the exception struct so that the ex handler can change the contents to be loaded back. I started to believe that is more useful now.

If we do that, we can remove this. We will only give write access to epc and fcr31 (and maybe sr and cr) so that the custom handler won't have to use inline assembly and at the same time it means we completely inverted the control to the consumer.

What do you think @rasky?

Reply:
I think that in general exception handlers might need to modify registers. It’s a pretty advanced use case, but why not. So maybe I’d simply drop const everywhere and document that it can be done.